File: 03_Stack/evaluateReversePolishNotation.go
package main

import (
	"fmt"
	"strconv"
)

func evalRPN(tokens []string) int {
	var stack []int
	var a, b int
	for _, c := range tokens {
		switch c {
		case "+":
			a, b, stack = getAndPopLastOperand(stack)
			stack = append(stack, (a + b))
		case "-":
			a, b, stack = getAndPopLastOperand(stack)
			stack = append(stack, (a - b))
		case "*":
			a, b, stack = getAndPopLastOperand(stack)
			stack = append(stack, (a * b))
		case "/":
			a, b, stack = getAndPopLastOperand(stack)
			stack = append(stack, (a / b))
		default:
			i, _ := strconv.Atoi(c)
			stack = append(stack, i)
		}
	}
	return stack[0]
}

func getAndPopLastOperand(stack []int) (int, int, []int) {
	a := stack[len(stack)-2]
	b := stack[len(stack)-1]
	stack = stack[:len(stack)-2]

	return a, b, stack
}

func main() {
	fmt.Println("Leetcode 150: Evaluate Reverse Polish Notation\n")

	// Example 1
	example1 := []string{"2", "1", "+", "3", "*"}
	fmt.Printf("Example 1: tokens = %v\n", example1)
	fmt.Println("Output", evalRPN(example1))

	// Example 2
	example2 := []string{"4", "13", "5", "/", "+"}
	fmt.Printf("Example 2: tokens = %v\n", example2)
	fmt.Println("Output", evalRPN(example2))

	// Example 3
	example3 := []string{"10", "6", "9", "3", "+", "-11", "*", "/", "*", "17", "+", "5", "+"}
	fmt.Printf("Example 3: tokens = %v\n", example3)
	fmt.Println("Output", evalRPN(example3))
}

File: 03_Stack/evaluate_reverse_polish_notation.py
class Solution:
    def evalRPN(self, tokens: list[str]) -> int:
        stack = []
        # loop through list
        for c in tokens:
            if c == "+":
                # pop last values add result to stack
                stack.append(stack.pop() + stack.pop())
            elif c == "-":
                # pop last values in list
                a, b = stack.pop(), stack.pop()
                # add calculation to stack
                stack.append(b - a)
            elif c == "*":
                stack.append(stack.pop() * stack.pop())
            elif c == "/":
                a, b = stack.pop(), stack.pop()
                stack.append(int(b / a))
            else:
                # if not an operator its a number so 
                # convert str to int
                stack.append(int(c))
        return stack[0]
    
# implementation
answer = Solution()

print("Leetcode 150: Evaluate Reverse Polish Notation\n")

# Example 1
eaxmple_1 = ["2", "1", "+", "3", "*"]
print(f"Example 1: tokens = {eaxmple_1}")
print("Output:", answer.evalRPN(eaxmple_1))

# Example 2
example_2 = ["4", "13", "5", "/", "+"]
print(f"Example 2: tokens = {example_2}")
print("Output:", answer.evalRPN(example_2))

# Example 3
example_3 = ["10", "6", "9", "3", "+", "-11", "*", "/", "*", "17", "+", "5", "+"]
print(f"Example 3: tokens = {example_3}")
print("Output:", answer.evalRPN(example_3))